### ISSUE TITLE
sympy struggles to evaluate complicated complex expressions

### ISSUE THREAD

**Issue by events555** [events555/CHP_AVX512#1]
# Introduction
To start off, some complex numbers are not returning their **arg**, that is, the angle the complex number makes with the real axis bounded between $(-\pi, \pi]$. This has been inconsistently reproduced, as inside the `discard_global_phase_state(mat)` **arg** is called upon the first non-zero element it finds, and in one case, it was the complex number $1/2 + j/2$. When testing this in an isolated command, `print((arg((1/2 + I/2))))` returns $0.785398163397448$ while `print((arg((1 + I)))` returns the expected $\frac{\pi}{4}$. Nevertheless, the code somehow calculates the $\frac{\pi}{4}$ and keeps the lingering complex exponentials.

This was observed with larger combinations of gates where the number of symbolic terms grew without simplification, as shown in the following section.

## Equivalent expressions not being evaluated
Another issue is that equivalent complex numbers are not being reduced and considered equal. Also note the complex exponentials obtained from the previous issue. Take as example the two expressions,
$$(1/4 - j/4)\cdot e^{j\frac{\pi}{4}} \text{ and } \frac{\sqrt{2}}{4}$$
If you were to check on Wolfram-Alpha, the first should also evaluate to $\frac{\sqrt{2}}{4}$, but it is also apparent as $1/4-j/4$ is also $\frac{\pi}{4}j$ but in the fourth quadrant.

As another example, here is one of the complicated outputs the program is trying to check if it is equal to an already discovered state.
$$\left(\frac{\sqrt{2}}{4} - \frac{\sqrt{2}i}{4}\right)e^{i\frac{\pi}{4}} + \left(-\frac{\sqrt{2}}{4} - \frac{\sqrt{2}i}{4}\right)\frac{e^{i\frac{\pi}{4}}}{2} + \left(\frac{\sqrt{2}}{4} + \frac{\sqrt{2}i}{4}\right)\frac{e^{i\frac{\pi}{4}}}{2} = 0.5$$
[This ](https://www.wolframalpha.com/input?i2d=true&i=%5C%2840%29Divide%5Bsqrt%5C%2840%292%5C%2841%29%2C4%5D+-+sqrt%5C%2840%292%5C%2841%29*Divide%5BI%2C4%5D%5C%2841%29*exp%5C%2840%29I*Divide%5Bpi%2C4%5D%5C%2841%29+%2B+%5C%2840%29-Divide%5Bsqrt%5C%2840%292%5C%2841%29%2C4%5D+-+sqrt%5C%2840%292%5C%2841%29*Divide%5BI%2C4%5D%5C%2841%29*Divide%5Bexp%5C%2840%29I*Divide%5Bpi%2C4%5D%5C%2841%29%2C2%5D+%2B+%5C%2840%29Divide%5Bsqrt%5C%2840%292%5C%2841%29%2C4%5D+%2B+sqrt%5C%2840%292%5C%2841%29*Divide%5BI%2C4%5D%5C%2841%29*Divide%5Bexp%5C%2840%29I*Divide%5Bpi%2C4%5D%5C%2841%29%2C2%5D)is the WolframAlpha evaluation which is equal to 0.5.

Beforehand, I was doing what Adeeb had done, where I was evaluating for the closeness of vectors after converting them to floating points. This had its own issues where sympy kept some symbolic factors, such as $\sqrt{2}$ while evaluating others as floating points, and the hashing function obviously treated them as separate.

As a final piece of information, it is known that the expected number of n-qubit stabilizer states is given as $2^n\prod_i (2^{i} + 1)$ from $i$ to $n$ as per this [answer](https://quantumcomputing.stackexchange.com/questions/11771/how-many-n-qubit-stabilizer-states-are-there). For 2 qubits and 3 qubits it is 60 and 1080 stabilizer states respectively. 

# Reproduction
```py
def discard_global_phase_state(mat):
    mat = mat.as_mutable()  # Convert to mutable matrix
    global_phase = None
    for i in range(mat.rows):
        if mat[i] != 0:  # Avoid division by zero
            global_phase = arg(mat[i])
            break
    if global_phase is not None:
        # Discard the global phase from all elements
        for i in range(mat.rows):
            mat[i] = mat[i] * exp(-I * global_phase)
    return mat.as_immutable()  # Convert back to immutable matrix if needed
```
Honestly, I don't know if this function accomplishes what I intended it to do. I know that when I evaluated the subsequent output as floats, it properly discarded equivalent stabilizer states for 2 qubits in 2 dimensions, but it was not working for 3 qubits and greater due to floating point error. I only got around this by discarding any values in the matrix elements less than 1e-10. 

Here is an example of what it is intended to do.
```py
    test1 = Matrix([[1/2 + I/2], [1/2 - I/2]])
    test2 = Matrix([[sqrt(2)/2], [sqrt(2)*I/2]])
    relPhase_test1 = discard_global_phase_state(test1)
    relPhase_test1_num = relPhase_test1.evalf()
    print(relPhase_test1) # prints Matrix([[(0.5 + I/2)*exp(-0.785398163397448*I)], [(0.5 - I/2)*exp(-0.785398163397448*I)]])

    print(relPhase_test1.evalf()) # prints Matrix([[0.707106781186548 + 2.1656938395398e-17*I], [2.1656938395398e-17 - 0.707106781186548*I]])
    print(test2.evalf()) # prints Matrix([[0.707106781186548], [0.707106781186548*I]])

    print(relPhase_test1 == test2) # prints false

```

These states should be equal because they only differ by some global phase.

# Desired Outcome
I believe the best way going forward is to figure out how to have sympy reduce the symbolic form because I was having trouble accumulating floating point errors on very large gate combinations. However, it might be true that only a few rounds are needed to hit all stabilizer states.

You can see in the above code that only after discarded the floating point values will `relPhase_test1_num` evaluate to equivalent to `test2.evalf()`. For reference, evaluating just the original test1 as a float gets you the matrix `Matrix([[0.5 + 0.5*I], [0.5 - 0.5*I]])` which is unhelpful.

**Comment by events555** [comment]
# Current Solution

I do not think this is the best method, but I believe it works in the general case.

For future reference, this function takes a complex column vector and calculates the angle that the first non-zero element makes with the real axis. It then divides this global phase from all other elements, essentially standardizing the phase across vectors.

```py
def discard_global_phase_state(mat):
    mat = mat.as_mutable()  # Convert to mutable matrix
    global_phase = None
    for i in range(mat.rows):
        if N(mat[i], 15, chop=True) != 0:  # Avoid division by zero, values close to 0 were getting through != and arg was returning NaN for 0
            global_phase = arg(mat[i])
            break
    if global_phase is not None:
        # Discard the global phase from all elements
        for i in range(mat.rows):
            mat[i] = mat[i] * exp(-I * global_phase)
    mat_num = N(mat, 15, chop=True)
    mat_simplify = mat_num.applyfunc(nsimplify)
    return mat_simplify.as_immutable()  # Convert back to immutable matrix
```

The main fixes I used for the aforementioned problems were

- I am numerically evaluating every element when checking for whether it is close to 0 
- After applying the global phase, I am evaluating it numerically one last time. This is to remove all extraneous symbols.
- I then use sympy's `nsimplify` function to attempt to return to a standard symbolic form that can be hashed.

The first point was because I was having trouble with arg returning NaN when complex numbers arbitrarily close to zero were not being evaluated as zero.

The second point was because I was having trouble with equivalent expressions not being evaluated as such due to sympy leaving many extra symbols instead of simplifying. 

The last point was because if I left it as the floating point evaluation, even after chopping there was sometimes lingering terms arbitrarily close to zero (terms on the order of 1e-17). **For some reason, nsimplify manages to get rid of these terms.**
